### DISCUSSION TITLE
npx setup-msw

### DISCUSSION THREAD

**Discussion by SevenOutman** [mswjs/msw#1548] [Show and tell]
Setup MSW with this single command

    npx setup-msw

![Screencast](https://github.com/SevenOutman/setup-msw/raw/main/screencast.svg)

Learn more at https://github.com/SevenOutman/setup-msw

**Comment by kettanaito** [comment]
That is _so_ cool! Thanks for building that, @SevenOutman! 

Would you be interested in discussing how we can improve this prompt and move it to be an official MSW starter? 

Some of the things from the top of my head:

- It should be possible to select multiple API kinds from the prompt (e.g. both REST and GraphQL in the same project). 
- We could generate dummy request handlers for the selected API kind to help people get started. A single `handlers.js` is a good place for that. 
- Similar to the API kinds, it should be possible to choose multiple environments, including both browser and Node.js. We could then create `browser.js` and `server.js` integrations accordingly.
- We could detect TypeScript and write generated files with the `.ts` extension.
- We could detect Jest/Vitest/Mocha test frameworks by their config files and propose to create a setup integration for them (this would be mind-blowing). 

What do you think about those?

**Reply by SevenOutman** [reply]
Hi @kettanaito , 

I appreciate it that you have interest in this and the offering of making it official! I'm really willing to. I already have answers for some of your interested topics:

> It should be possible to select multiple API kinds from the prompt (e.g. both REST and GraphQL in the same project).

The `setup-msw` script is simply running the official [Getting Started](https://mswjs.io/docs/getting-started) guide faithfully. It would definitely allowing selecting multiple API kinds if it's an officially suggested practice. Please let me know how you suggest the `handler.js` should be formed when user wants to declare both API kinds.

> We could generate dummy request handlers for the selected API kind to help people get started. A single handlers.js is a good place for that

It does generates them! The templates are based on https://mswjs.io/docs/getting-started/mocks/rest-api#response-resolver and https://mswjs.io/docs/getting-started/mocks/graphql-api#response-resolver

> Similar to the API kinds, it should be possible to choose multiple environments, including both browser and Node.js. We could then create browser.js and server.js integrations accordingly.

It makes sense but these two files don't seem to have difference in content at the setting-up stage. It might confuse user why it generates two duplicate files. WDYT?

> We could detect TypeScript and write generated files with the .ts extension.

It's doing it right now by detecting `tsconfig.json` presence. :)

> We could detect Jest/Vitest/Mocha test frameworks by their config files and propose to create a setup integration for them (this would be mind-blowing).

That's cool and I see it as the right next thing to do! The next minor release (in alpha) is already suggesting public directory path based on detected app framework (`./static` for Gatsby, `./src` for Angular, etc.), and I believe it would be so smooth if user can setup everything with no action but hitting Enter.

**Reply by kettanaito** [reply]
Sounds great!

> Please let me know how you suggest the handler.js should be formed when user wants to declare both API kinds.

Sure. The idea is that each API kind corresponds to 1 (?) example handler:

- REST: `rest.get('/resource', (req, res, ctx) => { ... })`
- GraphQL: `graphql.query('Resource', (req, res, ctx) => {})`

If you select multiple API kinds you get both examples merged in a single array of handlers in `handlers.js`:

```js
// handlers.js
import { rest, graphql } from 'msw'

export const handlers = [
  rest.get('/resource', (req, res, ctx) => { ... }),
  graphql.query('Resource', (req, res, ctx) => {})
]
```

> It makes sense but these two files don't seem to have difference in content at the setting-up stage. It might confuse user why it generates two duplicate files. WDYT?

They do, they use different `setup*` APIs:

```js
// browser.js
import { setupWorker } from 'msw'
import { handlers } from './handlers'

export const worker = setupWorker(...handlers)
```

```js
// server.js
import { setupServer } from 'msw/node'
import { handlers } from './handlers'

export const server = setupServer(...handlers)
```

I wouldn't actually start the worker/server in these files because they are meant to be integrated into the user's app:

- `worker.start()` must be called once before the app mounts;
- `server.listen()` is, ideally, integrated into your test setup and is also called once.

> It's doing it right now by detecting tsconfig.json presence. :)

Perfect!

**Reply by SevenOutman** [reply]
> The idea is that each API kind corresponds to 1 (?) example handler:
> If you select multiple API kinds you get both examples merged in a single array of handlers in `handlers.js`:

That makes sense. For now I used the example handlers from the docs, which are:

- **REST:** `.post('/login')` and `.get('/user')`
- **GraphQL:** `.mutation('/Login')` and `.query('GetUserInfo')`

Merging all 4 of them would result in a 70 line `handlers.js`. I think I should pick only 2 of them to not scare users.

> They do, they use different setup* APIs

I see. No idea why I missed that. 🤣

**Comment by SevenOutman** [comment]
[`setup-msw` v1.1.0](https://github.com/SevenOutman/setup-msw/releases/tag/v1.1.0) is release now, featuring:

- Suggest public directory based on the framework you use
- Selecting both mock API kinds
- Selecting both browser and Node.js integrations
